### PR TITLE
smartcontract: allow to use `*big.Int` numbers for integers

### LINKED ISSUE #2350: Allow to use big numbers in RPC queries
There are 2 issues which prevent us from using big integers when working via RPC:
1. Smartcontract `Integer` parameter expects `int64` as a value. https://github.com/nspcc-dev/neo-go/blob/master/pkg/smartcontract/parameter.go#L67 .
2. Request parameter is converted to `int` (not even `int64`) when building invocation script https://github.com/nspcc-dev/neo-go/blob/01d15ff473f9984023e9c1e04be107981acec80e/pkg/rpc/request/txBuilder.go#L65-66 .

### PR DESCRIPTION
**Pull request by fyrchik**
Close #2350 

Signed-off-by: Evgeniy Stratonikov <evgeniy@nspcc.ru>

### CODE REVIEW

File: pkg/smartcontract/convertor.go [hunk 1]
```diff
@@ -13,9 +13,14 @@ func ParameterFromStackItem(i stackitem.Item, seen map[stackitem.Item]bool) Para
...
 	case *stackitem.BigInteger:
+		bi := i.Value().(*big.Int)
+		var value interface{} = bi
+		if bi.IsInt64() {
+			value = bi.Int64()
+		}
 		return Parameter{
 			Type:  IntegerType,
-			Value: i.Value().(*big.Int).Int64(),
+			Value: value,
```
Comment (roman-khimov) [inline pkg/smartcontract/convertor.go:23]: Are there any users of `ParameterFromStackItem` at all?

File: pkg/smartcontract/parameter.go [hunk 2]
```diff
@@ -205,7 +225,11 @@ func (p *Parameter) EncodeBinary(w *io.BinWriter) {
 	case StringType:
 		w.WriteString(p.Value.(string))
 	case IntegerType:
-		w.WriteU64LE(uint64(p.Value.(int64)))
+		val, ok := p.Value.(*big.Int)
```
Comment (roman-khimov) [inline pkg/smartcontract/parameter.go:228]: Who is encoding `smartcontract.Parameter` into binary?

File: pkg/smartcontract/parameter.go [hunk 3]
```diff
@@ -236,7 +260,16 @@ func (p *Parameter) DecodeBinary(r *io.BinReader) {
 	case StringType:
 		p.Value = r.ReadString()
 	case IntegerType:
-		p.Value = int64(r.ReadU64LE())
+		bs := r.ReadVarBytes(bigint.MaxBytesLen)
```
Comment (roman-khimov) [inline pkg/smartcontract/parameter.go:263]: Where do we decode binary Parameter?

File: pkg/smartcontract/parameter.go [hunk 4]
```diff
@@ -152,10 +163,19 @@ func (p *Parameter) UnmarshalJSON(data []byte) (err error) {
...
 		}
-		if i, err = strconv.ParseInt(s, 10, 64); err != nil {
+		i, err = strconv.ParseInt(s, 10, 64)
+		if err == nil {
+			p.Value = i
 			return
 		}
-		p.Value = i
+
+		bi, ok := new(big.Int).SetString(s, 10)
```
Comment (roman-khimov) [inline pkg/smartcontract/parameter.go:172]: Maybe we can use `big.Int` always? It's only used by RPC client and CLI, it's not performance-critical.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

**Review by AnnaShaleva — COMMENTED** [review]
We also have `adjustValToType` that is currently able to work with int64 only, I think it should also be fixed (it is mosetly used by our CLI).

**Review by AnnaShaleva — COMMENTED** [review]
Also, we have a code inside our RPC client that works with int64 numbers (i.e. NEP17-related). Is it related to this PR?

File: pkg/smartcontract/parameter.go (hunk 2 above)
Comment (fyrchik) [reply pkg/smartcontract/parameter.go:228]: Nobody, it seems.

File: pkg/smartcontract/param_type.go [hunk 5]
```diff
@@ -250,8 +254,8 @@ func adjustValToType(typ ParamType, val string) (interface{}, error) {
 func inferParamType(val string) ParamType {
 	var err error
 
-	_, err = strconv.Atoi(val)
-	if err == nil {
+	_, ok := new(big.Int).SetString(val, 10)
```
Comment (roman-khimov) [inline pkg/smartcontract/param_type.go:257]: I think we need to `CheckIntegerSize` here.

File: pkg/smartcontract/param_type.go [hunk 6]
```diff
@@ -208,7 +208,11 @@ func adjustValToType(typ ParamType, val string) (interface{}, error) {
 			return nil, errors.New("invalid boolean value")
 		}
 	case IntegerType:
-		return strconv.ParseInt(val, 10, 64)
+		bi, ok := new(big.Int).SetString(val, 10)
```
Comment (roman-khimov) [inline pkg/smartcontract/param_type.go:211]: And here.

File: pkg/smartcontract/parameter.go [hunk 7]
```diff
@@ -145,17 +149,20 @@ func (p *Parameter) UnmarshalJSON(data []byte) (err error) {
...
 		}
 		// sometimes integer comes as string
-		if err = json.Unmarshal(r.Value, &s); err != nil {
-			return
+		if jErr := json.Unmarshal(r.Value, &s); jErr != nil {
+			return jErr
 		}
-		if i, err = strconv.ParseInt(s, 10, 64); err != nil {
-			return
+		bi, ok := new(big.Int).SetString(s, 10)
```
Comment (roman-khimov) [inline pkg/smartcontract/parameter.go:159]: And here.

File: pkg/smartcontract/parameter.go [hunk 8]
```diff
@@ -65,13 +65,17 @@ func (p Parameter) MarshalJSON() ([]byte, error) {
...
 	case IntegerType:
-		val, ok := p.Value.(int64)
+		val, ok := p.Value.(*big.Int)
 		if !ok {
 			resultErr = errors.New("invalid integer value")
 			break
 		}
-		valStr := strconv.FormatInt(val, 10)
-		resultRawValue = json.RawMessage(`"` + valStr + `"`)
+		err := stackitem.CheckIntegerSize(val)
```
Comment (roman-khimov) [inline pkg/smartcontract/parameter.go:73]: But do not do it here. We're marshaling the thing we have, whatever that is.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**